From 05895caf252ffb78940dafe6dc06b8989f48fc54 Mon Sep 17 00:00:00 2001
From: minindupasan <minindupasan@gmail.com>
Date: Mon, 28 Sep 2026 21:47:13 +0530
Subject: [PATCH 3/3] arm64: dts: p3768: add single IMX708 overlay for CAM1

The dual-camera overlay registers sensors on both CSI ports. With a
single camera fitted, the probe on the empty port fails with I2C
errors (-121) and can interfere with the fitted one. Add an overlay
that only describes the camera on CAM1 (serial_b, cam_i2cmux i2c@1),
selectable in jetson-io as "Camera IMX708-C".
---
 .../nvidia/t23x/nv-public/overlay/Makefile    |   1 +
 ...tegra234-p3768-camera-rpicam3-imx708-C.dts | 318 ++++++++++++++++++
 2 files changed, 319 insertions(+)
 create mode 100644 hardware/nvidia/t23x/nv-public/overlay/tegra234-p3768-camera-rpicam3-imx708-C.dts

diff --git a/hardware/nvidia/t23x/nv-public/overlay/Makefile b/hardware/nvidia/t23x/nv-public/overlay/Makefile
index 0c5b253..1aba7a3 100644
--- a/hardware/nvidia/t23x/nv-public/overlay/Makefile
+++ b/hardware/nvidia/t23x/nv-public/overlay/Makefile
@@ -76,6 +76,7 @@ dtbo-y += tegra234-p3767-camera-p3768-imx477-C.dtbo
 dtbo-y += tegra234-p3767-camera-p3768-imx477-A.dtbo
 dtbo-y += tegra234-p3740-0002-p3701-0008-csi.dtbo
 dtbo-y += tegra234-p3768-camera-rpicam3-imx708.dtbo
+dtbo-y += tegra234-p3768-camera-rpicam3-imx708-C.dtbo
 
 ifneq ($(dtb-y),)
 dtb-y := $(addprefix $(makefile-path)/,$(dtb-y))
diff --git a/hardware/nvidia/t23x/nv-public/overlay/tegra234-p3768-camera-rpicam3-imx708-C.dts b/hardware/nvidia/t23x/nv-public/overlay/tegra234-p3768-camera-rpicam3-imx708-C.dts
new file mode 100644
index 0000000..11331a7
--- /dev/null
+++ b/hardware/nvidia/t23x/nv-public/overlay/tegra234-p3768-camera-rpicam3-imx708-C.dts
@@ -0,0 +1,318 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+ * Copyright (c) 2023, RidgeRun.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/dts-v1/;
+/plugin/;
+
+
+#define CAM0_RST        TEGRA234_MAIN_GPIO(H, 3)
+#define CAM0_PWDN	TEGRA234_MAIN_GPIO(H, 6)
+#define CAM1_PWDN	TEGRA234_MAIN_GPIO(AC, 0)
+#define CAM_I2C_MUX 	TEGRA234_AON_GPIO(CC, 3)
+
+#include <dt-bindings/tegra234-p3767-0000-common.h>
+
+/ {
+	overlay-name = "Camera IMX708-C";
+	jetson-header-name = "Jetson 22pin CSI Connector";
+	compatible = JETSON_COMPATIBLE_P3768;
+
+	/* Single IMX708 on CAM1 connector (CSI port C, cam_i2cmux i2c@1) */
+	fragment@0 {
+		target-path = "/";
+		__overlay__ {
+			tegra-capture-vi {
+				num-channels = <1>;
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					port@0 {
+						reg = <0>;
+						rpicam3_imx708_vi_in1: endpoint {
+							status = "okay";
+							port-index = <2>;
+							bus-width = <2>;
+							remote-endpoint = <&rpicam3_imx708_csi_out1>;
+						};
+					};
+				};
+			};
+			tegra-camera-platform {
+				compatible = "nvidia, tegra-camera-platform";
+				modules {
+					module1 {
+						status = "okay";
+						badge = "jakku_rear_SC0872";
+						position = "rear";
+						orientation = "1";
+						drivernode0 {
+							status = "okay";
+							pcl_id = "v4l2_sensor";
+							sysfs-device-tree = "/sys/firmware/devicetree/base/bus@0/cam_i2cmux/i2c@1/rpicam3_imx708_c@1a";
+						};
+					};
+				};
+			};
+			bus@0 {
+				host1x@13e00000 {
+					nvcsi@15a00000 {
+						num-channels = <1>;
+						#address-cells = <1>;
+						#size-cells = <0>;
+						channel@0 {
+							reg = <0>;
+							ports {
+								#address-cells = <1>;
+								#size-cells = <0>;
+								port@0 {
+									reg = <0>;
+									rpicam3_imx708_csi_in1: endpoint@0 {
+										status = "okay";
+										port-index = <2>;
+										bus-width = <2>;
+										remote-endpoint = <&rpicam3_imx708_out1>;
+									};
+								};
+								port@1 {
+									reg = <1>;
+									rpicam3_imx708_csi_out1: endpoint@1 {
+										status = "okay";
+										remote-endpoint = <&rpicam3_imx708_vi_in1>;
+									};
+								};
+							};
+						};
+					};
+				};
+				cam_i2cmux {
+					status = "okay";
+					compatible = "i2c-mux-gpio";
+					#address-cells = <1>;
+					#size-cells = <0>;
+					mux-gpios = <&gpio_aon CAM_I2C_MUX GPIO_ACTIVE_HIGH>;
+					i2c-parent = <&cam_i2c>;
+					i2c@1 {
+						status = "okay";
+						reg = <1>;
+						#address-cells = <1>;
+						#size-cells = <0>;
+						imx708_cam1: rpicam3_imx708_c@1a {
+							status = "okay";
+							reset-gpios = <&gpio CAM1_PWDN GPIO_ACTIVE_HIGH>;
+							compatible = "sony,imx708";
+							reg = <0x1a>;
+							devnode = "video0";
+							physical_w = "3.680";
+							physical_h = "2.760";
+							sensor_model = "imx708";
+							use_sensor_mode_id = "true";
+
+							/* IMX708_MODE_4608x2592_14fps */
+							mode0 {
+								mclk_khz = "24000";
+								num_lanes = "2";
+								tegra_sinterface = "serial_c";
+								phy_mode = "DPHY";
+								discontinuous_clk = "yes";
+								dpcm_enable = "false";
+								cil_settletime = "0";
+								lane_polarity = "0";
+
+								active_w = "4608";
+								active_h = "2592";
+								mode_type = "bayer";
+								pixel_phase = "rggb";
+								csi_pixel_bit_depth = "10";
+								readout_orientation = "90";
+								line_length = "7824";
+								inherent_gain = "1";
+								mclk_multiplier = "80";
+								pix_clk_hz = "297600000";
+
+								gain_factor = "16";
+								framerate_factor = "1000000";
+								exposure_factor = "1000000";
+								min_gain_val = "16";
+								max_gain_val = "256";
+								step_gain_val = "1";
+								default_gain = "16";
+								min_hdr_ratio = "1";
+								max_hdr_ratio = "1";
+								min_framerate = "2000000";
+								max_framerate = "14000000";
+								step_framerate = "1";
+								default_framerate = "14000000";
+								min_exp_time = "500";
+								max_exp_time = "65487";
+								step_exp_time = "1";
+								default_exp_time = "1600";
+								embedded_metadata_height = "4";
+							};
+							/* IMX708_MODE_2304x1296_56fps (2x2 binned) */
+							mode1 {
+								mclk_khz = "24000";
+								num_lanes = "2";
+								tegra_sinterface = "serial_c";
+								phy_mode = "DPHY";
+								discontinuous_clk = "yes";
+								dpcm_enable = "false";
+								cil_settletime = "0";
+								lane_polarity = "0";
+
+								active_w = "2304";
+								active_h = "1296";
+								mode_type = "bayer";
+								pixel_phase = "rggb";
+								csi_pixel_bit_depth = "10";
+								readout_orientation = "90";
+								line_length = "3976";
+								inherent_gain = "1";
+								mclk_multiplier = "80";
+								pix_clk_hz = "297600000";
+
+								gain_factor = "16";
+								framerate_factor = "1000000";
+								exposure_factor = "1000000";
+								min_gain_val = "16";
+								max_gain_val = "256";
+								step_gain_val = "1";
+								default_gain = "16";
+								min_hdr_ratio = "1";
+								max_hdr_ratio = "1";
+								min_framerate = "2000000";
+								max_framerate = "56000000";
+								step_framerate = "1";
+								default_framerate = "56000000";
+								min_exp_time = "500";
+								max_exp_time = "65487";
+								step_exp_time = "1";
+								default_exp_time = "1600";
+								embedded_metadata_height = "4";
+							};
+							/* IMX708_MODE_1536x864_120fps (2x2 binned, centre crop) */
+							mode2 {
+								mclk_khz = "24000";
+								num_lanes = "2";
+								tegra_sinterface = "serial_c";
+								phy_mode = "DPHY";
+								discontinuous_clk = "yes";
+								dpcm_enable = "false";
+								cil_settletime = "0";
+								lane_polarity = "0";
+
+								active_w = "1536";
+								active_h = "864";
+								mode_type = "bayer";
+								pixel_phase = "rggb";
+								csi_pixel_bit_depth = "10";
+								readout_orientation = "90";
+								line_length = "2741";
+								inherent_gain = "1";
+								mclk_multiplier = "80";
+								pix_clk_hz = "297600000";
+
+								gain_factor = "16";
+								framerate_factor = "1000000";
+								exposure_factor = "1000000";
+								min_gain_val = "16";
+								max_gain_val = "256";
+								step_gain_val = "1";
+								default_gain = "16";
+								min_hdr_ratio = "1";
+								max_hdr_ratio = "1";
+								min_framerate = "2000000";
+								max_framerate = "120000000";
+								step_framerate = "1";
+								default_framerate = "120000000";
+								min_exp_time = "500";
+								max_exp_time = "65487";
+								step_exp_time = "1";
+								default_exp_time = "1600";
+								embedded_metadata_height = "4";
+							};
+							/* IMX708_MODE_2304x1296_HDR_30fps (sensor HDR, 2x2 downscaled) */
+							mode3 {
+								mclk_khz = "24000";
+								num_lanes = "2";
+								tegra_sinterface = "serial_c";
+								phy_mode = "DPHY";
+								discontinuous_clk = "yes";
+								dpcm_enable = "false";
+								cil_settletime = "0";
+								lane_polarity = "0";
+
+								active_w = "2304";
+								active_h = "1296";
+								mode_type = "bayer";
+								pixel_phase = "rggb";
+								csi_pixel_bit_depth = "10";
+								readout_orientation = "90";
+								line_length = "1996";
+								inherent_gain = "1";
+								mclk_multiplier = "80";
+								pix_clk_hz = "297600000";
+
+								gain_factor = "16";
+								framerate_factor = "1000000";
+								exposure_factor = "1000000";
+								min_gain_val = "16";
+								max_gain_val = "256";
+								step_gain_val = "1";
+								default_gain = "16";
+								min_hdr_ratio = "1";
+								max_hdr_ratio = "1";
+								min_framerate = "2000000";
+								max_framerate = "30000000";
+								step_framerate = "1";
+								default_framerate = "30000000";
+								min_exp_time = "500";
+								max_exp_time = "65487";
+								step_exp_time = "1";
+								default_exp_time = "1600";
+								embedded_metadata_height = "4";
+							};
+
+							ports {
+								#address-cells = <1>;
+								#size-cells = <0>;
+
+								port@0 {
+									reg = <0>;
+									rpicam3_imx708_out1: endpoint {
+										status = "okay";
+										port-index = <2>;
+										bus-width = <2>;
+										remote-endpoint = <&rpicam3_imx708_csi_in1>;
+									};
+								};
+							};
+						};
+					};
+				};
+				gpio@2200000 {
+					camera-control-output-low {
+						gpio-hog;
+						output-low;
+						gpios = <CAM0_PWDN 0>;
+						label = "cam0-pwdn";
+					};
+				};
+			};
+		};
+	};
+};
